### ISSUE TITLE
[semi] Enforce semi after interface declaration

### ISSUE THREAD

**Issue by OoDeLally** [typescript-eslint/typescript-eslint#1444]
**Repro**

Set `.eslintrc.js` following https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/semi.md

```js
{
  "semi": "off",
  "@typescript-eslint/semi": ["error"]
}
```

Then declare an interface in a `ts` file.
```ts
interface Foo {
  ...
}
```

**Expected Result**

```ts
interface Foo {
  ...
} // Error: Missing semicolumn.
```

**Actual Result**

```ts
interface Foo {
  ...
} // No error.
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.15.0` |
| `@typescript-eslint/parser`        | `2.15.0` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `v8.16.0` |
| `npm`                              | `6.4.1` |

**Comment by armano2** [comment]
> See also the @typescript-eslint/member-delimiter-style rule, which allows you to specify the delimiter for type and interface members.

**Comment by bradzacher** [comment]
Why do you want semicolons after interface declarations? As Armano mentioned - this is invalid syntax, and will result in extraneous semicolons being omitted in your code.

interfaces fall into a category of things called declarations. These things do not need semicolons, as they have a well defined end - the final `}`.

Examples of other things that are declarations - functions, classes, enums, namespaces.

```ts
interface IFace { };
enum Enum { };
class Class { };
function func() { };
namespace Namespace { };
namespace Namespace { export const x = 1 };

//////////////////// transpiles to: ////////////////////

"use strict";
;
var Enum;
(function (Enum) {
})(Enum || (Enum = {}));
;
class Class {
}
;
function func() { }
;
;
var Namespace;
(function (Namespace) {
    Namespace.x = 1;
})(Namespace || (Namespace = {}));
;

```
[repl](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgJIDFEoN7IL4DcAUBCAK4C2yAouVboUQgDZwDObyAwqx8g8RhkQCMMAD2IZEJEAKAJT98xEHAoQ2AByzIAcmo3akSxqvVad+80ZzIIAD03ioYZAkltX95AF5kARmUiIA)

Notice in the above example how the output looks really weird, with a bunch of semicolons on new lines? That's because the semicolons aren't supposed to be included.

As such, we won't be adding an option to enforce incorrect syntax.

**Comment by OoDeLally** [comment]
This is fair, however then why is it enforced for `type`?

![image](https://user-images.githubusercontent.com/1717629/72300184-4a093100-3663-11ea-8385-0b4ae78efe86.png)

I am not sure I understand the difference of treatments between `type` and `interface`, since it seems to me that they are both declarations.

**Comment by armano2** [comment]
because typescript allows to use semicolon after `type` type alias declaration, its something similar to variable declaration but for types.

https://www.typescriptlang.org/play/#code/C4TwDgpgBAZg9nKBeKAmA3EA

------------

In JS you should not put semicolon after block like nodes 
```
for (a of b) {}
```
```
function x () {}
```
```
class X { }
```
------------------

for other nodes you can, and in some cases have to put semi after some statement.

for example, code like this is going to work in typescript
```
interface x {} interface Y {} interface Z {} // single line
```
but code like this will not
```
type x = number type Y = string type Z = any // single line
````

**Comment by bradzacher** [comment]
You're certainly not wrong! They are both declarations.

The rules are not as clear cut as I made them out to be. Essentially, every single thing in JS(/TS) has a "grammar" associated with it.

For things like functions, classes, enums, and namespaces; the grammar is well defined such that they have a clearly defined termination point, which is the closing `}`.
Because the grammar has a clearly defined termination point, you don't need a semicolon to delimit the end of the statement.

For things like variable declarations (`const x = 1`), and type declarations (`type T = string`), the grammar is a lot more ambiguous.
Their grammar is more like this:
`<keyword> <whitespace or newline> <identifier> <whitespace or newline> = <whitespace or newline> <expression>;`
The important bit is that after the equals, they have an `<expression>`.
An expression is very abstract, and [encompasses many different language constructs](https://github.com/typescript-eslint/typescript-eslint/blob/b5ef704d2620acac5dbfa8bc21a859b898d61286/packages/typescript-estree/src/ts-estree/ts-estree.ts#L285-L303), which means that it's not well defined where the termination point is.

Because it's not well defined, you need a delimiter to signify the end of the statement, which is the semicolon.

You've got some examples of when delimiters help - when things are on a single line.
But there's also this really good example:
```ts
const x = someVariable
(x as () => void)();
```

In this example, how is this parsed? eyeballing it, it looks like it should be:
```ts
const x = someVariable // end of statement
(x as () => void)(); // end of statement
```
However in reality, it is parsed like this:
```ts
const x = someVariable(x as () => void)();
```

Why? Because of the ambiguity of the expression grammar!
In particular, the call expression grammar is specified like this:
`<expression> <whitespace or newline> ( <whitespace or newline> <arguments...> <whitespace or newline> ) `
Without a semicolon, the parser doesn't know that the `(` at the start of the next line is supposed to be a new statement.

In practice, the types grammar is a lot tighter, so you don't run into the same issues; but to be sure you don't run into ambiguous cases they included the semicolon in the grammar.

---

That was a bit long - so TL;DR is that it depends on how each thing is defined to be parsed by the language.
